src/backend/utils/cache/plancache.c

```diff
@@ -856,6 +860,41 @@ CheckCachedPlan(CachedPlanSource *plansource)
 	return false;
 }
 
+/*
+ * detect_plangen_switch: check if plangen switched between statements.
+ *
+ * Check if plangen switched for statements of one plan or between plans.
+ * 'plan' should point to currently initialized plan.
+ * 'plansource' is where plangen summary stored for all plans.
+ */
+static void
+detect_plangen_switch(CachedPlan *plan, CachedPlanSource *plansource)
+{
+	if (plansource->plangen_switched)
+		return;
+
+	ListCell   *lc;
+	bool		first_stmt = true;
```

Comment:
It is possible to get rid of `first_stmt` flag: use save  `planGen` from `lfirst(plan->stmt_list)` and then in `for_each_cell(lc, lnext(plan->stmt_list)` compare it with `planGen` of others.

Reply:
No need to save it in that case. Just change `if (first_stmt && plansource->generation == 1)` to `if (lfirst(plan->stmt_list) == plannedstmt && plansource->generation == 1)`.

Reply:
Like this:

```
static void
detect_plangen_switch(CachedPlan *plan, CachedPlanSource *plansource)
{
	ListCell   *lc;
	PlannedStmt *plannedstmt;
	
	if (plansource->plangen_switched)
		return;

	plannedstmt = (PlannedStmt *) lfirst(plan->stmt_list);
	
	if (plannedstmt && plansource->generation == 1)
	{
		/* Save very first statement's plangen value ... */
		plansource->init_plangen_used = plannedstmt->planGen;
	}
	
	/* ... and compare all others with it at next iteration or call. */
	for_each_cell(lc, lnext(plan->stmt_list))
	{
		plannedstmt = (PlannedStmt *) lfirst(lc);
		if (plansource->init_plangen_used != plannedstmt->planGen)
		{
			plansource->plangen_switched = true;
			break;
		}
	}
}

```

Reply:
```C

	plannedstmt = (PlannedStmt *) lfirst(plan->stmt_list);
	
	if (plannedstmt && plansource->generation == 1)
	{
		/* Save very first statement's plangen value ... */
		plansource->init_plangen_used = plannedstmt->planGen;
	}
	
	/* ... and compare all others with it at next iteration or call. */
	for_each_cell(lc, lnext(plan->stmt_list))
	{
                ...
```
Here we miss processing of first statement on second and further plan generation when `plansource->generation > 1`.

Reply:
@seryozhasmirnov I would change your fragment as follows:
```C
static void
detect_plangen_switch(CachedPlan *plan, CachedPlanSource *plansource)
{
	ListCell   *lc;
	PlannedStmt *plannedstmt;
	
	if (plansource->plangen_switched)
		return;

        assert(list_length(plan->stmt_list) > 0);
	plannedstmt = (PlannedStmt *) linitial(plan->stmt_list);
        assert(plannedstmt);
	
	if (plansource->generation == 1)
	{
		/* Save very first statement's plangen value ... */
		plansource->init_plangen_used = plannedstmt->planGen;
	} else if (plansource->init_plangen_used != plannedstmt->planGen)
                plansource->plangen_switched = true;
        }
	
	/* ... and compare all others with it at next iteration or call. */
	for_each_cell(lc, lnext(plan->stmt_list))
	{
		plannedstmt = (PlannedStmt *) lfirst(lc);
		if (plansource->init_plangen_used != plannedstmt->planGen)
		{
			plansource->plangen_switched = true;
			break;
		}
	}
}
```

Reply:
@seryozhasmirnov @maksm90 
I suggest the following version of `detect_plangen_switch` if we really need it.
```C
static void
detect_plangen_switch(CachedPlan *plan, CachedPlanSource *plansource)
{
	Assert(list_length(plan->stmt_list) > 0);

	ListCell	*lc = list_head(plan->stmt_list);
	PlannedStmt	*plannedstmt;

	if (plansource->plangen_switched)
		return;

	/* Save very first statement's plangen value ... */
	if (plansource->generation == 1)
	{
		plannedstmt = (PlannedStmt *) lfirst(lc);
		plansource->init_plangen_used = plannedstmt->planGen;
		lc = lnext(lc);
	}
	/* ... and compare all others with it at next iteration or call. */
	for_each_cell(lc, lc)
	{
		plannedstmt = (PlannedStmt *) lfirst(lc);
		if (plansource->init_plangen_used != plannedstmt->planGen)
		{
			plansource->plangen_switched = true;
			break;
		}
	}
}
```

Reply:
>  I suggest the following version of detect_plangen_switch if we really need it.

LGTM

> 	Assert(list_length(plan->stmt_list) > 0);

**Please notice**: assert is not correct here - you have to explicitly handle the case when `stmt_list` is empty. This case might occur if we pass `;` statement using extended query protocol. For example, the following C client program emulates this scenario:
```C
// run as `<exec name> <connection string>`

#include <assert.h>
#include <stdio.h>
#include "libpq-fe.h"

int main(int argc, char **argv) {
    assert(argc == 2);

    const char *conninfo = argv[1];
    PGconn     *conn = PQconnectdb(conninfo);
    if (PQstatus(conn) != CONNECTION_OK)
    {
        fprintf(stderr, "Connection to database failed: %s",
                PQerrorMessage(conn));
        return 1;
    }

    PGresult *res = PQexecParams(conn, ";", 0, NULL, NULL, NULL, NULL, 1);
    PQclear(res);
    PQfinish(conn);
    printf("Done\n");

    return 0;
}
```

Reply:
@maksm90 Case approved. Thanks.